include/deal.II/base/mpi.h

```diff
@@ -268,6 +272,71 @@ namespace Utilities
       MPI_Comm comm;
     };
 
+    /**
+     * An implementation of a critical section that guards an algorithm with MPI
+     * communication when executed several times.
+     *
+     * This critical section enforces that a section of code (between the
+     * constructor and destructor of this helper class) only starts executing
+     * after all MPI ranks in the same communicator finished execution of this
+     * section from an earlier time.
+     *
+     * The class internally executes a non-blocking barrier in the destructor
+     * for each critical section and waits for the completion of the barrier
+     * when constructed.
+     *
+     * This makes certain MPI algorithms correct that rely on receiving messages
+     * using MPI_ANY_SOURCE, where executing the same algorithm more than once
+     * in a row can confuse messages between the first and second execution.
+     *
+     * Example usage:
+     * @code
+     * {
+     *   static MPI_Request request = MPI_REQUEST_NULL;
+     *   CriticalSection cs(comm, request);
+     *
+     *   // [code to be guarded]
+     * }
+     * @endcode
+     */
+    class CriticalSection
```

Comment:
This is definitely a critical section for us. If we consider two following accesses to the same section as *processes* and the MPI library as a *shared recourse*, we could indeed call it `CriticalSection`.  But nevertheless, I would not use the term *critical section* here since if I would see this in a code, I would think the access to a shared resource (like a file) are synchronized.

What we are rather doing with this class, is to mark a region (of non-collective MPI calls) as a user-defined collective (MPI) operation, whose correctness is not guaranteed if two calls are launched simultaneously...

This is why I would propose to use something like `MPI::UserCollectiveOperation` or `MPI::CriticalCollectiveOperation`. @tjhei what is your opinion?